[net] Various fixes to ftpd, ktcp, and TIME_WAIT

ghaerr opened the pull request:
Various fixes discussed in #999 and elsewhere.

Summary:

- ftpd: add code to become a daemon, add to build, distribution and net start/stop.
- ktcp: fix FIN_WAIT2 bug when FIN and ACK received in FIN_WAIT_1; now goes directly to TIME_WAIT state.
- All timeouts now have 1/16 second granularity.
- TIME_WAIT default to half second.

I'm currently unable to test ELKS ftpd as I probably have the wrong settings for ftp'ing into QEMU.

ghaerr committed:
[net] Various fixes to ftpd, ktcp, and TIME_WAIT

Mellvik commented:
Thanks @ghaerr, that was fast indeed!

And thanks again for the TIME_WAIT links - very enlightening.

Looking forward to test in the morning.

--Mellvik


> 11. nov. 2021 kl. 21:05 skrev Gregory Haerr ***@***.***>:
> 
> ﻿
> Various fixes discussed in #999 and elsewhere.
> 
> Summary:
> 
> ftpd: add code to become a daemon, add to build, distribution and net start/stop.
> ktcp: fix FIN_WAIT2 bug when FIN and ACK received in FIN_WAIT_1; now goes directly to TIME_WAIT state.
> All timeouts now have 1/16 second granularity.
> TIME_WAIT default to half second.
> I'm currently unable to test ELKS ftpd as I probably have the wrong settings for ftp'ing into QEMU.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/jbruchon/elks/pull/1001
> 
> Commit Summary
> 
> [net] Various fixes to ftpd, ktcp, and TIME_WAIT
> File Changes  (7 files)
> M elkscmd/Applications (3)
> M elkscmd/inet/Makefile (1)
> M elkscmd/inet/ftp/ftpd.c (16)
> M elkscmd/ktcp/tcp.c (8)
> M elkscmd/ktcp/tcp.h (21)
> M elkscmd/ktcp/tcp_cb.c (4)
> M elkscmd/rootfs_template/bin/net (4)
> Patch Links:
> 
> https://github.com/jbruchon/elks/pull/1001.patch
> https://github.com/jbruchon/elks/pull/1001.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

ghaerr committed:
Fix in_gethostbyname, urlget/ftpget error messages, time_wait to 4 secs

ghaerr committed:
Small fix to force short jumps in unreal.S

ghaerr commented:
@Mellvik,

Pushed more fixes:

- Fixed in_gethostbyname, was always return last entry when host not found in /etc/hosts
- Better error message in net_connect in ftpget for when hostname not found
- Changed ftpget/ftpput to return actual program name in error messages instead of always 'ftp:' for clarity
- Reset TIME_WAIT to 4 seconds (instead of 1/2 second), still down from 10 seconds

Now that I realize what is happening the last ^P debug trace now makes sense: when the following line was being seen
```
tcp: port 20 reused, other socket NOT in time_wait!
```
I didn't at first understand it. The reason is, the bug @Mellvik found where the closed socket was going into FIN_WAIT_2 state, rather than TIME_WAIT, and so previous socket alloc was not being released when port 20 was being reused.

Now that this PR should fix that problem, I don't think there's a need to radically reduce TIME_WAIT, since the ftp daemon will cause that socket control block to be freed when it reuses the port. This will in turn free up the memory that is causing ktcp to run out of heap. The message we should be seeing if working properly is:
```
"tcp: port 20 reused, freeing previous socket in time_wait
```
This PR does reduce the TIME_WAIT to 4 seconds, which should help a bit more to release memory, while not interfering in possible close problems (until we finally figure them all out!!).

Since I can't (yet) access ftpd outside of ELKS (in QEMU), I started trying to use ftpget to access ftpd. It is always failing with 'ftpget: Error listing directory: 502`. I haven't jumped into this, but was wondering if you could fix that? That would be very helpful!

Thank you!

Mellvik commented:
This is great, @ghaerr. 

I'll be back shortly with test results. I created a collection of about 50 small files to transfer - that will be the litmus test for the new socket resource release regime. 
Now - from one of the TIME_WAIT links you sent yesterday, it is likely to fail eventually - which is OK. It's more like a feature, not a bug. And it's our privilgege to adjust the timeout to make sense - in our setting. Ideally, there would be an under the hood, automatic, per connection round trip time recording which would set the timeout. I don't think that exists anyware.

Will look into the urlget/ftpd problem right away. Are you using urlget on the mac?

--Mellvik


> 12. nov. 2021 kl. 00:33 skrev Gregory Haerr ***@***.***>:
> 
> ﻿
> @Mellvik,
> 
> Pushed more fixes:
> 
> Fixed in_gethostbyname, was always return last entry when host not found in /etc/hosts
> Better error message in net_connect in ftpget for when hostname not found
> Changed ftpget/ftpput to return actual program name in error messages instead of always 'ftp:' for clarity
> Reset TIME_WAIT to 4 seconds (instead of 1/2 second), still down from 10 seconds
> Now that I realize what is happening the last ^P debug trace now makes sense: when the following line was being seen
> 
> tcp: port 20 reused, other socket NOT in time_wait!
> I didn't at first understand it. The reason is, the bug @Mellvik found where the closed socket was going into FIN_WAIT_2 state, rather than TIME_WAIT, and so previous socket alloc was not being released when port 20 was being reused.
> 
> Now that this PR should fix that problem, I don't think there's a need to radically reduce TIME_WAIT, since the ftp daemon will cause that socket control block to be freed when it reuses the port. This will in turn free up the memory that is causing ktcp to run out of heap. The message we should be seeing if working properly is:
> 
> "tcp: port 20 reused, freeing previous socket in time_wait
> This PR does reduce the TIME_WAIT to 4 seconds, which should help a bit more to release memory, while not interfering in possible close problems (until we finally figure them all out!!).
> 
> Since I can't (yet) access ftpd outside of ELKS (in QEMU), I started trying to use ftpget to access ftpd. It is always failing with 'ftpget: Error listing directory: 502`. I haven't jumped into this, but was wondering if you could fix that? That would be very helpful!
> 
> Thank you!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.
